homeassistant/components/fireservicerota/__init__.py

```diff
@@ -0,0 +1,182 @@
+
+        self.availability_data = None
+        self.incident_data = None
+        self.incident_id = None
+        self.response_data = None
+
+        self.fsr_avail = FireServiceRota(
+            base_url=f"https://{self._url}", token_info=self._tokens
+        )
+
+        self.fsr_incidents = FireServiceRotaIncidents(on_incident=self.on_incident)
+
+        self.start_listener()
+
+    def construct_url(self) -> str:
+        """Return url with latest config values."""
+        return WSS_BWRURL.format(
+            self._entry.data[CONF_URL], self._entry.data[CONF_TOKEN]["access_token"]
+        )
+
+    def on_incident(self, data) -> None:
+        """Update the current data."""
+        _LOGGER.debug("Got data from websocket listener: %s", data)
+        self.incident_data = data
+
+        async_dispatcher_send(self._hass, f"{DOMAIN}_{self._entry.entry_id}_update")
+
+    def start_listener(self) -> None:
+        """Start the websocket listener."""
+        _LOGGER.debug("Starting incidents listener")
+        self.fsr_incidents.start(self.construct_url())
+
+    def stop_listener(self) -> None:
+        """Stop the websocket listener."""
+        _LOGGER.debug("Stopping incidents listener")
+        self.fsr_incidents.stop()
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    async def async_update(self) -> None:
+        """Get the latest availability data."""
+        try:
+            self.availability_data = await self._hass.async_add_executor_job(
+                self.fsr_avail.get_availability, str(self._hass.config.time_zone)
+            )
+            _LOGGER.debug("Updating availability data")
+        except (ExpiredTokenError, InvalidTokenError):
+            _LOGGER.debug("Refreshing expired tokens")
+            await self.async_refresh_tokens()
+
+    async def async_response_update(self) -> None:
+        """Get the latest incident response data."""
+        if self.incident_data:
+            self.incident_id = self.incident_data.get("id")
+            _LOGGER.debug("Incident id: %s", self.incident_id)
+            try:
+                self.response_data = await self._hass.async_add_executor_job(
+                    self.fsr_avail.get_incident_response, self.incident_id
+                )
+                _LOGGER.debug("Updating incident response data")
+            except (ExpiredTokenError, InvalidTokenError):
+                _LOGGER.debug("Refreshing expired tokens")
+                await self.async_refresh_tokens()
+
+    async def async_set_response(self, incident_id, value) -> None:
+        """Set incident response status."""
+        try:
+            await self._hass.async_add_executor_job(
+                self.fsr_avail.set_incident_response, incident_id, value
+            )
+            _LOGGER.debug("Setting incident response status")
+        except (ExpiredTokenError, InvalidTokenError):
+            _LOGGER.debug("Refreshing expired tokens")
+            await self.async_refresh_tokens()
```

Comment:
This logic is repeating. Make this a decorator, context manager or create a helper function

Reply:
Created helper function update_call(), code becomes smaller and smaller ;-)